macros/parserNumberWithUnits.pl

```diff
@@ -37,11 +37,46 @@ =head1 DESCRIPTION
 We now call on the Legacy version, which is used by
 num_cmp to handle numbers with units.
 
+New units can be added at run time by using the newUnit option
+
+       $a = NumberWithUnits("3 apples",{newUnit=>'apples'});
+
+A new unit can either be a string, in which case the string is added as a
+new unit with no relation to other units, or as a hashreference
+
+       $newUnit = {name => 'bear',
+                   conversion => {factor =>3, m=>1}};
+       $a = NumberWithUnits("3 bear", {newUnit=>$newUnit});
+
+You can also define your own conversion hash.  In the above example one bear
+is three meters.  (See Units.pm for examples). 
+
+Finally, the newUnit option can also be an array ref containing any number of
+new units to add.  A common reason for doing this would be to add the plural
+version of the unit as an equilvalent unit.  E.G.
+
+      $newUnits = ['apple',{name=>'apples',conversion=>{factor=>1,apple=>1}}];
+      $a = NumberWithUnits("3 apples",{newUnit=>$newUnits});
+
+In this case both 3 apple and 3 apples would be considered correct.  
+
+Note:  English pluralization is suprisingly hard, so WeBWorK will make no 
+attempt to display a grammerically correct result.  
+
 =cut
 
 loadMacros('MathObjects.pl');
 
 sub _parserNumberWithUnits_init {
+  # We make copies of these hashes here because these copies will be unique to  # the problem.  The hashes in Units are shared between problems.  We pass
+  # the hashes for these local copies to the NumberWithUnits package to use
+  # for all of its stuff.  
+  %fundamental_units = %Units::fundamental_units;
+  %known_units = %Units::known_units;
```

Comment:
I didn't mean to add the ~ files.  They don't actually work right now, but
I had to update my personal repo so I could transfer between machines and
those changes showed up here. Its what I get for working on something in
consideration for pull requests.  I'll get rid of them.

Its not always clear to me why some variables are local and some are global
in PG.  For example %Units::known_units is available as a variable
directly.  If I add `my`'s should I also add accessors? Its not impossible
that someone would want to be able to see what units are available.  If I
do leave them as global then I could also give them better names.

On Sun, Jan 31, 2016 at 9:52 AM, Davide P. Cervone <notifications@github.com
> wrote:

> In macros/parserNumberWithUnits.pl
> <https://github.com/openwebwork/pg/pull/268#discussion_r51361440>:
>
> >  =cut
> >
> >  loadMacros('MathObjects.pl');
> >
> >  sub _parserNumberWithUnits_init {
> > +  # We make copies of these hashes here because these copies will be unique to  # the problem.  The hashes in Units are shared between problems.  We pass
> > +  # the hashes for these local copies to the NumberWithUnits package to use
> > +  # for all of its stuff.
> > +  %fundamental_units = %Units::fundamental_units;
> > +  %known_units = %Units::known_units;
>
> Same here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openwebwork/pg/pull/268/files#r51361440>.
>